examples/flutter_gallery/lib/demo/material/slider_demo.dart

```diff
@@ -11,12 +13,124 @@ class SliderDemo extends StatefulWidget {
   _SliderDemoState createState() => new _SliderDemoState();
 }
 
+Path _triangle(double size, Offset thumbCenter, {bool invert: false}) {
+  final Path thumbPath = new Path();
+  final double height = math.sqrt(3.0) / 2.0;
+  final double halfSide = size / 2.0;
+  final double centerHeight = size * height / 3.0;
+  final double sign = invert ? -1.0 : 1.0;
+  thumbPath.moveTo(thumbCenter.dx - halfSide, thumbCenter.dy + sign * centerHeight);
+  thumbPath.lineTo(thumbCenter.dx, thumbCenter.dy - 2.0 * sign * centerHeight);
+  thumbPath.lineTo(thumbCenter.dx + halfSide, thumbCenter.dy + sign * centerHeight);
+  thumbPath.close();
+  return thumbPath;
+}
+
+class _CustomThumbShape extends SliderComponentShape {
+  const _CustomThumbShape();
+  static const double _thumbSize = 4.0;
+  static const double _disabledThumbSize = 4.0;
+
+  @override
+  Size getPreferredSize(bool isEnabled, bool isDiscrete) {
+    return new Size.fromRadius(isEnabled ? _thumbSize : _disabledThumbSize);
+  }
+
+  @override
+  void paint(
+    RenderBox parentBox,
+    PaintingContext context,
+    bool isDiscrete,
+    Offset thumbCenter,
+    Animation<double> activationAnimation,
+    Animation<double> enableAnimation,
+    TextPainter labelPainter,
+    SliderThemeData sliderTheme,
+    TextDirection textDirection,
+    double value,
```

Comment:
looking at this API I wonder if it would make sense to wrap all these values into a single immutable object and pass that in, instead. It would make it much easier for us to add fields later without breaking people.

Reply:
Hmm. Good point.  I'll try that out.

Reply:
OK, I looked at doing that, but the problem is that you still have to recreate the tweens, etc. every paint.

So, I looked at just making a bunch of no-op setters on the abstract interface.  That way if you care about it, you implement the setter, and if you don't, then you don't implement it.  You can also optimize the setter to only do work if the value changed from last time. The slider calls all of the setters each time it paints.

We can still extend it with new fields, because each one has a default no-op implementation and won't break anyone who doesn't care about it.

I left a few things in the argument list that seemed more like ephemeral things, but I could move them out too (paintContext, thumbCenter, and value).